plugin/src/compilation/decompiler.ml

```diff
@@ -70,63 +101,101 @@ let get_pushed_names env =
             failwith "Unexpected Anonymous in get_pushed_names."
          | Name n -> n) names)
 
+(* Checks if we are applying function g, which
+   is expected to have len arguments. *)
+let try_fun f args g len : (constr array) option =
```

Comment:
When it comes to supporting particular functions like this, I prefer the approach we take in the library a lot using custom datatypes so that we don't have to hardcode in argument numbers anywhere except in one centralized place, and records assign meaning to each argument. Like we do in [sigmautils](https://github.com/uwplse/coq-plugin-lib/blob/master/src/coq/constants/sigmautils.ml), [equtils](https://github.com/uwplse/coq-plugin-lib/blob/master/src/coq/constants/sigmautils.ml), and so on. So I would also do that with `or_introl` and so on in proputils in coq-plugin-lib. This function, though, could still be useful when constructing those, so you may still want to use it within proputils---if so, you can put it in `apputils`.